File: leetcode/78_Subsets.py
import unittest
from typing import List

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        def rec(i: int, nums: List[int], combination: List[int], combinations: List[List[int]]) -> None:
            if i >= len(nums):
                combinations.append(combination.copy())
            else:
                rec(i + 1, nums, combination, combinations)
                combination.append(nums[i])
                rec(i + 1, nums, combination, combinations)
                combination.pop()
        combinations = []
        rec(0, nums, [], combinations)
        return combinations

class TestSubsets(unittest.TestCase):
    def test_1(self):
        self.assertEqual(sorted(Solution().subsets([1,2,3])), sorted([[],[1],[2],[1,2],[3],[1,3],[2,3],[1,2,3]]))    

    def test_2(self):
        self.assertEqual(Solution().subsets([0]), [[],[0]])    

File: leetcode/78_Subsets.rs
struct Solution {}

impl Solution {
    pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
        fn rec(i: usize, nums: &Vec<i32>, combination: &mut Vec<i32>, combinations: &mut Vec<Vec<i32>>) {
            if i >= nums.len() {
                combinations.push(combination.clone());
            } else {
                rec(i + 1, nums, combination, combinations);
                combination.push(nums[i]);
                rec(i + 1, nums, combination, combinations);
                combination.pop();
            }
        }
        let mut combinations: Vec<Vec<i32>> = vec![];
        rec(0, &nums, &mut vec![], &mut combinations);
        combinations       
    }
}

#[cfg(test)]
mod sort_colors_test {
    use super::*;

    #[test]
    fn test_1() {
        let mut result = Solution::subsets(vec![1,2,3]);
        let mut target = vec![vec![], vec![1], vec![2], vec![1,2], vec![3], vec![1,3], vec![2,3], vec![1,2,3]];
        result.sort();
        target.sort();
        assert_eq!(result, target);
    }    

    #[test]
    fn test_2() {
        assert_eq!(Solution::subsets(vec![0]), vec![vec![], vec![0]]);
    }    
}